packages/beacon-node/src/chain/historicalState/utils/blockReplay.ts

```diff
@@ -0,0 +1,101 @@
+import {PubkeyIndexMap} from "@chainsafe/pubkey-index-map";
+import {BeaconConfig} from "@lodestar/config";
+import {
+  BeaconStateAllForks,
+  CachedBeaconStateAllForks,
+  DataAvailableStatus,
+  ExecutionPayloadStatus,
+  createCachedBeaconState,
+  stateTransition,
+} from "@lodestar/state-transition";
+import {Slot} from "@lodestar/types";
+import {IBeaconDb} from "../../../db/index.js";
+import {HistoricalStateRegenMetrics, RegenErrorType} from "../types.js";
+import {measure} from "@lodestar/utils";
+
+/**
+ * Populate a PubkeyIndexMap with any new entries based on a BeaconState
+ */
+export function syncPubkeyCache(state: BeaconStateAllForks, pubkey2index: PubkeyIndexMap): void {
+  // Get the validators sub tree once for all the loop
+  const validators = state.validators;
+
+  const newCount = state.validators.length;
+  for (let i = pubkey2index.size; i < newCount; i++) {
+    const pubkey = validators.getReadonly(i).pubkey;
+    pubkey2index.set(pubkey, i);
+  }
+}
+
+/**
+ * Get and regenerate a historical state
+ */
+export async function replayBlocks(
+  {
+    toSlot,
+    lastFullSlot,
+    lastFullStateBytes,
+  }: {
+    toSlot: Slot;
+    lastFullStateBytes: Uint8Array;
+    lastFullSlot: Slot;
+  },
+  {
+    config,
+    db,
+    pubkey2index,
+    metrics,
+  }: {config: BeaconConfig; db: IBeaconDb; pubkey2index: PubkeyIndexMap; metrics?: HistoricalStateRegenMetrics}
+): Promise<Uint8Array> {
+  if (lastFullSlot + 1 !== toSlot) {
+    throw new Error(`Invalid full state slot to regen historical sate. expected=${toSlot - 1} actual=${lastFullSlot}`);
+  }
+
+  return measure(metrics?.stateTransitionTime, async () => {
+    let state = config.getForkTypes(toSlot).BeaconState.deserializeToViewDU(lastFullStateBytes);
+    syncPubkeyCache(state, pubkey2index);
+    state = createCachedBeaconState(
+      state,
+      {
+        config,
+        pubkey2index,
+        index2pubkey: [],
+      },
+      {
+        skipSyncPubkeys: true,
+      }
+    );
+
+    let blockCount = 0;
+    for await (const block of db.blockArchive.valuesStream({gt: lastFullSlot, lte: toSlot})) {
+      try {
+        state = stateTransition(
+          state as CachedBeaconStateAllForks,
+          block,
+          {
+            verifyProposer: false,
+            verifySignatures: false,
+            verifyStateRoot: false,
+            executionPayloadStatus: ExecutionPayloadStatus.valid,
+            dataAvailableStatus: DataAvailableStatus.available,
+          },
+          metrics
+        );
+      } catch (e) {
+        metrics?.regenErrorCount.inc({reason: RegenErrorType.blockProcessing});
+        throw e;
+      }
+      blockCount++;
+      if (Buffer.compare(state.hashTreeRoot(), block.message.stateRoot) !== 0) {
```

Comment:
do this inside the for loop for every block because if it happens there it will not reach here